gen/suites/actor_deletion/main.go

```diff
@@ -0,0 +1,136 @@
+		// 		ID:      "fail-delete-w-balance-and-unkown-beneficiary",
+		// 		Version: "v1",
+		// 		Desc:    "fails when actor with non-zero balance is deleted but beneficiary address is unknown",
+		// 	},
+		// 	Selector: Selector{"chaos_actor": "true"},
+		// 	Func:     deleteActorWithBeneficiary(big.NewInt(50), MustNewSECP256K1Addr("!👹*_👹!"), exitcode.SysErrorIllegalActor),
+		// },
+		&MessageVectorGenItem{
+			Metadata: &Metadata{
+				ID:      "fail-delete-w-balance-and-self-beneficiary",
+				Version: "v1",
+				Desc:    "fails when actor with non-zero balance is deleted but beneficiary is the calling actor",
+				Comment: "should abort if the beneficiary is the calling actor, see https://github.com/filecoin-project/specs-actors/blob/bcd83e8eb0a98b684851e574a2bd8d4130e21a51/actors/runtime/runtime.go#L117",
+			},
+			Selector: Selector{"chaos_actor": "true"},
+			Hints:    []string{HintIncorrect, HintNegate},
+			Func:     deleteActorWithBeneficiary(big.NewInt(50), chaos.Address, exitcode.Ok),
+		},
+	)
+}
+
+// deleteActor builds a test vector that tests the simple case of deleting an
+// actor with zero balance.
+func deleteActor(v *Builder) {
+	v.Messages.SetDefaults(GasLimit(1_000_000_000), GasPremium(1), GasFeeCap(200))
+
+	sender := v.Actors.Account(address.SECP256K1, big.NewInt(1_000_000_000_000_000))
+	v.CommitPreconditions()
+
+	v.Assert.ActorExists(chaos.Address)
+	v.Assert.BalanceEq(chaos.Address, big.Zero())
+
+	delMsg := v.Messages.Raw(
+		sender.ID,
+		chaos.Address,
+		chaos.MethodDeleteActor,
+		MustSerialize(&builtin.BurntFundsActorAddr),
+		Value(big.Zero()),
+		Nonce(0),
+	)
+	v.CommitApplies()
+
+	v.Assert.NoError(ExitCode(exitcode.Ok)(delMsg.Result))
+	v.Assert.ActorMissing(chaos.Address)
+}
+
+// deleteActorWithBeneficiary builds a test vector that tests deleting an actor
+// and transfering their funds to another actor. Use address.Undef to have a
+// beneficiary created automatically. actorFunds MUST be greater than zero.
+func deleteActorWithBeneficiary(actorFunds big.Int, beneficiaryAddr address.Address, code exitcode.ExitCode) func(v *Builder) {
+	return func(v *Builder) {
+		v.Messages.SetDefaults(GasLimit(1_000_000_000), GasPremium(1), GasFeeCap(200))
+
+		sender := v.Actors.Account(address.SECP256K1, big.Add(big.NewInt(1_000_000_000_000_000), actorFunds))
+
+		if beneficiaryAddr == address.Undef {
+			beneficiaryAddr = v.Actors.Account(address.SECP256K1, big.Zero()).ID
+		}
+
+		v.CommitPreconditions()
+
+		v.Assert.ActorExists(chaos.Address)
+		v.Assert.BalanceEq(chaos.Address, big.Zero())
+
+		// transfer required funds to the actor that will be deleted
+		m := v.Messages.Sugar().Transfer(sender.ID, chaos.Address, Value(actorFunds), Nonce(0))
+		v.Messages.ApplyOne(m)
+		v.Assert.EveryMessageResultSatisfies(ExitCode(exitcode.Ok))
+
+		// if this is will succeed, record the current balance so we can check funds
+		// were transferred to the beneficiary
+		var bal big.Int
+		if code == exitcode.Ok {
```

Comment:
I think this if is superfluous (in general we shouldn't be having ifs in test vectors, rather asserts).

Reply:
Note that this is conditional operates on the `code` param (the expected exit code) not an actual exit code from an applied message. In some vectors we expect the exit code to not be `Ok` and there's a couple of things we can't/shoudln't do when it's not ok:

1. We can't retrieve the initial balance (since the actor may not exist)
2. We shouldn't check funds were transferred to the beneficiary because it shouldn't of happened

So I can split this into multiple functions to deal with success/failure exit codes but there will be a bunch of code repeted code. WDYT?

Reply:
Ah, I totally misread this. This makes sense, sorry. Just for clarity, can we change the parameter name to `expectedCode`? (Maybe we should do the same for all other test vectors that accept the expected code as a parameter; it's more verbose but clearer).

Reply:
Sure no problem 😄